core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -66,6 +68,11 @@ private[history] class FsHistoryProvider(conf: SparkConf) extends ApplicationHis
   @volatile private var applications: mutable.LinkedHashMap[String, FsApplicationHistoryInfo]
     = new mutable.LinkedHashMap()
 
+  // The logCheckingThread appends elements at the end of lazyApplications, while
+  // the logLazyReplayThread removes elements from the begin of lazyApplications.
+  private val lazyApplications = new mutable.LinkedHashMap[String, LazyFsApplicationHistoryInfo]()
+    with mutable.SynchronizedMap[String, LazyFsApplicationHistoryInfo]
```

Comment:
though here since we need to maintain the order, we might have to use `ConcurrentSkipListMap` instead